scipy/optimize/_milp.py

```diff
@@ -0,0 +1,359 @@
+                raise ValueError(message)
+        As.append(csc_array(constraint.A))
+        b_ls.append(np.atleast_1d(constraint.lb).astype(np.double))
+        b_us.append(np.atleast_1d(constraint.ub).astype(np.double))
+
+    if len(As) > 1:
+        A = vstack(As)
+        b_l = np.concatenate(b_ls)
+        b_u = np.concatenate(b_us)
+    else:  # avoid unnecessary copying
+        A = As[0]
+        b_l = b_ls[0]
+        b_u = b_us[0]
+
+    return A, b_l, b_u
+
+
+def _milp_iv(c, integrality, bounds, constraints, options):
+
+    # objective IV
+    c = np.atleast_1d(c).astype(np.double)
+    if c.ndim != 1:
+        message = "`c` must be a one-dimensional array."
+        raise ValueError(message)
+
+    # integrality IV
+    message = ("`integrality` must contain integers 0-3 and be broadcastable "
+               "to `c.shape`.")
+    if integrality is None:
+        integrality = 0
+    try:
+        integrality = np.broadcast_to(integrality, c.shape).astype(np.uint8)
+    except ValueError:
+        raise ValueError(message)
+    if integrality.min() < 0 or integrality.max() > 3:
+        raise ValueError(message)
+
+    # bounds IV
+    if bounds is None:
+        bounds = Bounds(0, np.inf)
+    elif not isinstance(bounds, Bounds):
+        message = "`bounds` must be an instance of `scipy.optimize.Bounds`."
+        try:
+            bounds = Bounds(*bounds)
+        except TypeError:
+            raise ValueError(message)
+
+    try:
+        lb = np.broadcast_to(bounds.lb, c.shape).astype(np.double)
+        ub = np.broadcast_to(bounds.ub, c.shape).astype(np.double)
+    except (ValueError, TypeError):
+        message = ("`bounds.lb` and `bounds.ub` must contain reals and "
+                   "be broadcastable to `c.shape`.")
+        raise ValueError(message)
+
+    # constraints IV
+    if not constraints:
+        constraints = [LinearConstraint(np.empty((0, c.size)),
+                                        np.empty((0,)), np.empty((0,)))]
+    A, b_l, b_u = _constraints_to_components(constraints)
+
+    if A.shape != (b_l.size, c.size):
+        message = "The shape of `A` must be (len(b_l), len(c))."
+        raise ValueError(message)
+    indptr, indices, data = A.indptr, A.indices, A.data.astype(np.double)
+
+    # options IV
+    options = options or {}
+    supported_options = {'disp', 'presolve', 'time_limit'}
+    unsupported_options = set(options).difference(supported_options)
+    if unsupported_options:
+        message = (f"Unrecognized options detected: {unsupported_options}. "
+                   "These will be passed to HiGHS verbatim.")
+        warnings.warn(message, RuntimeWarning, stacklevel=3)
+    options_iv = {'log_to_console': options.get("disp", False)}
+    options_iv.update(options)
+
+    return c, integrality, lb, ub, indptr, indices, data, b_l, b_u, options_iv
+
+
+def milp(c, *, integrality=None, bounds=None, constraints=None, options=None):
+    r"""
+    Mixed-integer linear programming
```

Comment:
The algorithm is deterministic right? I think it should be said somewhere.

Reply:
Added to the notes:
> The algorithm is deterministic, and it typically finds the global optimum of
    moderately challenging mixed-integer linear programs (when it exists).